stdlib/public/core/StringNormalization.swift

```diff
@@ -102,3 +102,253 @@ internal func _tryNormalize(
+  }
+  
+  var outputBufferThreshold: Int {
+    return outputBuffer.count - 4
+  }
+
+  // TODO: Additional fast-path: All CCC-ascending NFC_QC segments are NFC
+  // TODO: Just freakin do normalization and don't bother with ICU
+  var outputCount = 0
+  let outputEnd = outputBufferThreshold
+  var inputCount = 0
+  let inputEnd = sourceBuffer.count
+  while inputCount < inputEnd && outputCount < outputEnd {
+    // TODO: Slightly faster code-unit scan for latiny (<0xCC)
+
+    // Check scalar-based fast-paths
+    let (scalar, len) = _decodeScalar(sourceBuffer, startingAt: inputCount)
+    _internalInvariant(inputCount &+ len <= inputEnd)
+
+    if _slowPath(
+         !sourceBuffer.hasNormalizationBoundary(before: inputCount &+ len)
+      || !isNFCStarter(scalar)
+    ) {
+      break
+    }
+    inputCount &+= len
+
+    for cu in UTF8.encode(scalar)._unsafelyUnwrappedUnchecked {
+      outputBuffer[outputCount] = cu
+      outputCount &+= 1
+    }
+
+    _internalInvariant(inputCount == outputCount,
+      "non-normalizing UTF-8 fast path should be 1-to-1 in code units")
+  }
+  return (inputCount, outputCount)
+}
+
+internal func transcodeToUTF16(
+  _ sourceBuffer: UnsafeBufferPointer<UInt8>,
+  into outputBuffer: UnsafeMutableBufferPointer<UInt16>
+) -> (Int, Int)? {
+  var readIndex = 0
+  var writeIndex = 0
+  let outputCount = outputBuffer.count
+  let sourceCount = sourceBuffer.count
+  
+  while readIndex < sourceCount {
+    let (scalar, length) = _decodeScalar(sourceBuffer, startingAt: readIndex)
+    
+    if scalar._hasNormalizationBoundaryBefore && readIndex != 0 {
+      break
+    }
+    
+    readIndex += length
+    
+    for cu in scalar.utf16 {
+      if writeIndex < outputCount {
+        outputBuffer[writeIndex] = cu
+        writeIndex += 1
+      } else {
+        return nil
+      }
+    }
+  }
+  
+  return (readIndex, writeIndex)
+}
+
+internal func transcodeToUTF8(
+  _ sourceBuffer: UnsafeBufferPointer<UInt16>,
+  into outputBuffer: UnsafeMutableBufferPointer<UInt8>
+) -> (Int, Int)? {
+  var readIndex = 0
+  var writeIndex = 0
+  let outputCount = outputBuffer.count
+  let sourceCount = sourceBuffer.count
+  
+  while readIndex < sourceCount {
+    let (scalar, length) = _decodeScalar(sourceBuffer, startingAt: readIndex)
+    //we don't need to check for normalization boundaries here because we are only transcoding
+    //a single segment at this point
+    
+    readIndex += length
+    
+    for cu in UTF8.encode(scalar)._unsafelyUnwrappedUnchecked {
+      if writeIndex < outputCount {
+        outputBuffer[writeIndex] = cu
+        writeIndex &+= 1
+      } else {
+        return nil
+      }
+    }
+  }
+  return (readIndex, writeIndex)
+}
+
+internal func fastNormalize(
+  readIndex: String.Index,
+  guts: _StringGuts,
+  outputBuffer: UnsafeMutableBufferPointer<UInt8>,
+  icuInputBuffer: UnsafeMutableBufferPointer<UInt16>,
+  icuOutputBuffer: UnsafeMutableBufferPointer<UInt16>
+) -> NormalizationResult {
```

Comment:
Do we want to take `_guts` here? We'd prefer not to dispatch once in the caller rather than every time we call this function. `withFastUTF8` checks if small, and spills to the stack, then checks if shared and calls a function for that, and finally interprets as native.